### ISSUE TITLE
spec-interp validation differs from wasm-validate

### ISSUE THREAD

**Issue by sbc100** [WebAssembly/wabt#1176]
When a wasm file doesn't have a DataCount section I think validation of  memory.init should fail.  I have a wasm file that validates in wabt but in v8 reports:

`CompileError: WebAssembly.Module(): Compiling function #5840 failed: invalid data segment index: 2 @+2798285
`

**Comment by amirbawab** [comment]
Hi @sbc100, can you please post the wasm file?

**Comment by tlively** [comment]
Here's a trivial example.

```
(module
 (memory 1 1)
 (data passive "")
 (func (data.drop 0))
)
```
You can get a binary with no DataCount section using Binaryen's wasm-opt and passing the `--mvp-features --no-validation` flags.

**Comment by sbc100** [comment]
Presumably we are missing a spec test case here.

**Comment by sbc100** [comment]
Seems like there is a spec test for this

However we pass the test because we use binary-reader-interp.cc when loading all the spec tests and the interpreter has its own check for this.  However the check is missing from the validator itself.

**Comment by binji** [comment]
Probably should have a fuller solution to this (difference between spec-interp and wasm-validate), but this particular bug is fixed in #1297.

**Comment by binji** [comment]
spec-interp and wasm-validate now share a validator (`SharedValidator`), so they produce the same results.
